bionic/core/task_execution.py

```diff
@@ -580,9 +581,9 @@ def sync_after_remote_computation(self):
         # do anything if it fails now.)
         self._cache_accessor.flush_stored_entries()
 
-        # Then, populate the value hashes.
-        if self._result_value_hash is None:
-            self._load_value_hash()
+        artifact = self._get_artifact()
+        self._cache_accessor.save_artifact(artifact)
```

Comment:
Nice catch! 

@jqmp Although the fix works, I think it can easily happen again in the future. What do you both think about changing the persistence `load_artifact` API to save the artifact locally when downloading from a remote source? I mean we are already downloading the artifact from remote, might as well save it.

Reply:
I'm not sure I totally understand this bug -- as you mentioned elsewhere, it would be great for this PR to include a description of what's going wrong. But in theory the cache accessor should be storing a copy of any artifact object it creates. Is the problem that `accessor.load_artifact` is creating a local copy but not saving it? If so, I agree that we should probably ensure that calling `load_artifact` twice doesn't result in two downloads. Off the top of my head, I'm not sure if it makes more sense to actually save the local artifact, or just to maintain an in-memory copy of the local artifact object. (The former seems more robust, but also means that `load_artifact` can now mutate the cache state. Maybe that's not a big deal....)

As far as this happening in the future: I agree! Can we add a test that checks for multiple downloads? I see this PR updates the logging tests, but those don't seem like they'd catch duplicate messages.

Reply:
The logging tests were updated to check that the total count of the messages is the same as the expected. `assert len(actual_messages) == len(expected_patterns)`. But yes, I think a more robust double download check would be nicer.

Reply:
Ah, sorry, I missed that change! That makes sense.

As you said, I do think a more explicit check would be good, but I guess you'll be the best judge of how practical it is.